File: Backend/src/Fibertest30.Api/Mapping/RftsEventsMapping.cs
namespace Fibertest30.Api;

public static class RftsEventsMapping
{
    public static RftsEventsData ToProto(this Iit.Fibertest.Dto.RftsEventsDto dto)
    {
        var rftsEventsData = new RftsEventsData
        {
            ErrorMessage = dto.ErrorMessage ?? "", // Convert null to empty string
            IsNoFiber = dto.IsNoFiber,
            LevelArray = { dto.LevelArray?.Select(ToProto) ?? [] },
            Summary = dto.Summary != null ? ToProto(dto.Summary) : null
        };

        return rftsEventsData;
    }

    private static RftsLevel ToProto(this Iit.Fibertest.Dto.RftsLevelDto dto)
    {
        return new RftsLevel
        {
            Level = dto.Level.ToProto(),
            IsFailed = dto.IsFailed,
            FirstProblemLocation = dto.FirstProblemLocation ?? "",
            EventArray = { dto.EventArray.Select(e => e.ToProto()) },
            TotalFiberLoss = dto.TotalFiberLoss.ToProto(),
        };
    }

    private static RftsWords ToProto(this Iit.Fibertest.Dto.RftsWords word)
    {
        return word switch
        {
            Iit.Fibertest.Dto.RftsWords.yes => RftsWords.Yes,
            Iit.Fibertest.Dto.RftsWords.fail => RftsWords.Fail,
            Iit.Fibertest.Dto.RftsWords.pass => RftsWords.Pass,
            Iit.Fibertest.Dto.RftsWords.newEvent => RftsWords.NewEvent,
            Iit.Fibertest.Dto.RftsWords.fiberBreak => RftsWords.FiberBreak,
            Iit.Fibertest.Dto.RftsWords.empty => RftsWords.Empty,
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    private static RftsEvent ToProto(this Iit.Fibertest.Dto.RftsEventDto dto)
    {
        var rftsEvent = new RftsEvent
        {
            Ordinal = dto.Ordinal,
            IsNew = dto.IsNew,
            IsFailed = dto.IsFailed,
            LandmarkTitle = dto.LandmarkTitle,
            LandmarkType = dto.LandmarkType.ToProto(),
            State = dto.State.ToProto(),
            DamageType = dto.DamageType,
            DistanceKm = dto.DistanceKm,
            Enabled = dto.Enabled.ToProto(),
            EventType = dto.EventType,
            ReflectanceCoeff = dto.ReflectanceCoeff ?? "",
            AttenuationInClosure = dto.AttenuationInClosure ?? "",
            AttenuationCoeff = dto.AttenuationCoeff ?? "",
            ReflectanceCoeffDeviation = dto.ReflectanceCoeffDeviation ?? "",
            AttenuationInClosureDeviation = dto.AttenuationInClosureDeviation ?? "",
            AttenuationCoeffDeviation = dto.AttenuationCoeffDeviation ?? "",
        };

        if (dto.ReflectanceCoeffThreshold != null)
            rftsEvent.ReflectanceCoeffThreshold = dto.ReflectanceCoeffThreshold.ToProto();
        if (dto.AttenuationInClosureThreshold != null)
            rftsEvent.AttenuationInClosureThreshold = dto.AttenuationInClosureThreshold.ToProto();
        if (dto.AttenuationCoeffThreshold != null)
            rftsEvent.AttenuationCoeffThreshold = dto.AttenuationCoeffThreshold.ToProto();

        return rftsEvent;
    }

    private static TotalFiberLoss ToProto(this Iit.Fibertest.Dto.TotalFiberLossDto dto)
    {
        return new TotalFiberLoss
        {
            Value = dto.Value,
            Threshold = dto.Threshold.ToProto(),
            Deviation = dto.Deviation,
            IsPassed = dto.IsPassed,
        };
    }

    private static MonitoringThreshold ToProto(this Iit.Fibertest.Dto.MonitoringThreshold dto)
    {
        return new MonitoringThreshold
        {
            Value = dto.Value,
            IsAbsolute = dto.IsAbsolute,
        };

    }

    private static RftsEventsSummary ToProto(this Iit.Fibertest.Dto.RftsEventsSummaryDto dto)
    {
        return new RftsEventsSummary
        {
            TraceState = dto.TraceState.ToProto(),
            BreakLocation = dto.BreakLocation,
            Orl = dto.Orl,
            LevelStates = { dto.LevelStates?.Select(ls => ls.ToProto()) ?? [] },
        };
    }

    private static LevelState ToProto(this Iit.Fibertest.Dto.LevelState dto)
    {
        return new LevelState
        {
            LevelTitle = dto.LevelTitle,
            State = dto.State,
        };
    }
}

File: WebFt30/src/app/core/store/mapping/rfts-events-mapping.ts
import * as grpc from 'src/grpc-generated';
import {
  LevelState,
  MonitoringThreshold,
  RftsEvent,
  RftsEvents,
  RftsEventsSummary,
  RftsLevel,
  TotalFiberLoss
} from '../models/ft30/rfts-events-dto';
import { FtEnumsMapping } from './ft-enums-mapping';

export class RftsEventsMapping {
  static fromRftsEventsDto(grpcRftsEvents: grpc.RftsEventsData): RftsEvents {
    return {
      errorMessage: grpcRftsEvents.errorMessage,
      isNoFiber: grpcRftsEvents.isNoFiber,
      levelArray: grpcRftsEvents.levelArray?.map((level) => this.fromRftsLevelDto(level)) || [],
      summary: grpcRftsEvents.summary
        ? this.fromRftsEventsSummaryDto(grpcRftsEvents.summary)
        : new RftsEventsSummary()
    };
  }

  static fromRftsLevelDto(grpcLevel: grpc.RftsLevel): RftsLevel {
    return {
      level: FtEnumsMapping.fromGrpcFiberState(grpcLevel.level),
      isFailed: grpcLevel.isFailed,
      firstProblemLocation: grpcLevel.firstProblemLocation,
      eventArray: grpcLevel.eventArray?.map((event) => this.fromRftsEventDto(event)) || [],
      totalFiberLoss: grpcLevel.totalFiberLoss
        ? this.fromTotalFiberLossDto(grpcLevel.totalFiberLoss)
        : new TotalFiberLoss()
    };
  }

  static fromGrpcRftsWords(grpcRftsWord: grpc.RftsWords): string {
    // prettier-ignore
    switch(grpcRftsWord) {
        case grpc.RftsWords.Yes: return "i18n.ft.yes";
        case grpc.RftsWords.Fail: return "i18n.ft.fail";
        case grpc.RftsWords.Pass: return "i18n.ft.pass";
        case grpc.RftsWords.NewEvent: return "i18n.ft.new";
        case grpc.RftsWords.FiberBreak: return "i18n.ft.fiber-break";
        case grpc.RftsWords.Empty: return "";
        default:
          throw new Error(`Unknown grpc.RftsWords: ${grpcRftsWord}`);  }
  }

  static fromRftsEventDto(grpcEvent: grpc.RftsEvent): RftsEvent {
    return {
      ordinal: grpcEvent.ordinal,
      isNew: grpcEvent.isNew,
      isFailed: grpcEvent.isFailed,
      landmarkTitle: grpcEvent.landmarkTitle,
      landmarkType: grpcEvent.landmarkType,
      state: this.fromGrpcRftsWords(grpcEvent.state),
      damageType: grpcEvent.damageType,
      distanceKm: grpcEvent.distanceKm,
      enabled: this.fromGrpcRftsWords(grpcEvent.enabled),
      eventType: grpcEvent.eventType,
      reflectanceCoeff: grpcEvent.reflectanceCoeff,
      attenuationInClosure: grpcEvent.attenuationInClosure,
      attenuationCoeff: grpcEvent.attenuationCoeff,
      reflectanceCoeffThreshold: grpcEvent.reflectanceCoeffThreshold
        ? this.fromMonitoringThresholdDto(grpcEvent.reflectanceCoeffThreshold)
        : null,
      attenuationInClosureThreshold: grpcEvent.attenuationInClosureThreshold
        ? this.fromMonitoringThresholdDto(grpcEvent.attenuationInClosureThreshold)
        : null,
      attenuationCoeffThreshold: grpcEvent.attenuationCoeffThreshold
        ? this.fromMonitoringThresholdDto(grpcEvent.attenuationCoeffThreshold)
        : null,
      reflectanceCoeffDeviation: grpcEvent.reflectanceCoeffDeviation,
      attenuationInClosureDeviation: grpcEvent.attenuationInClosureDeviation,
      attenuationCoeffDeviation: grpcEvent.attenuationCoeffDeviation
    };
  }

  static fromTotalFiberLossDto(grpcTotalFiberLoss: grpc.TotalFiberLoss): TotalFiberLoss {
    return {
      value: grpcTotalFiberLoss.value,
      threshold: grpcTotalFiberLoss.threshold
        ? this.fromMonitoringThresholdDto(grpcTotalFiberLoss.threshold)
        : new MonitoringThreshold(),
      deviation: grpcTotalFiberLoss.deviation,
      isPassed: grpcTotalFiberLoss.isPassed
    };
  }

  static fromMonitoringThresholdDto(grpcThreshold: grpc.MonitoringThreshold): MonitoringThreshold {
    return {
      value: grpcThreshold.value,
      isAbsolute: grpcThreshold.isAbsolute
    };
  }

  static fromRftsEventsSummaryDto(grpcSummary: grpc.RftsEventsSummary): RftsEventsSummary {
    return {
      traceState: FtEnumsMapping.fromGrpcFiberState(grpcSummary.traceState),
      breakLocation: grpcSummary.breakLocation,
      orl: grpcSummary.orl,
      levelStates:
        grpcSummary.levelStates?.map((levelState) => this.fromLevelStateDto(levelState)) || []
    };
  }

  static fromLevelStateDto(grpcLevelState: grpc.LevelState): LevelState {
    return {
      levelTitle: grpcLevelState.levelTitle,
      state: grpcLevelState.state
    };
  }
}
